L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc

```diff
@@ -117,21 +121,89 @@ unsigned short HGCalTriggerTowerGeometryHelper::getTriggerTowerFromEtaPhi(const
-  return getTriggerTowerFromEtaPhi(thesum.position().eta(), thesum.position().phi());
+std::unordered_map<unsigned short, float> HGCalTriggerTowerGeometryHelper::getTriggerTower(const l1t::HGCalTriggerSums& thesum) const {
+  std::unordered_map<unsigned short, float> binIDandShares = {}; 
+  if(!splitModuleSum_){
+    binIDandShares.insert({getTriggerTowerFromEtaPhi(thesum.position().eta(), thesum.position().phi()), 1.0});
+    return binIDandShares;
+  } else{
+    HGCalModuleDetId detid(thesum.detId());
+    int moduleU = detid.moduleU();
+    int moduleV = detid.moduleV();
+    int layer = detid.layer();
+    int sector = detid.sector();
+    int zside = detid.zside();
+
+    int subdet = -1;
+    int splitDivisor = -1;
+    if(detid.isHScintillator()){ //FIXME HFNose not handled
+      subdet = 2;
+      layer += 28;
+      splitDivisor = splitDivisorScint_; 
+    } else if(detid.isEE()){
+      subdet = 0;
+      splitDivisor = splitDivisorSilic_;
+    } else if(detid.isHSilicon()) {
+      subdet = 1;
+      layer += 28;
+      splitDivisor = splitDivisorSilic_;
+    }
+    
+    std::ifstream moduleTowerMappingStream(moduleTowerMapping_.fullPath());
+    if (!moduleTowerMappingStream.is_open()) {
+      throw cms::Exception("MissingDataFile") << "Cannot open HGCalTowerMapProducer moduleTowerMapping file\n";
+    }
+
+    std::vector<std::string> result; // the line including towers and module sum shares for this module
+    std::string line;
+    getline(moduleTowerMappingStream, line); //To skip first row
+    for( std::string line; getline(moduleTowerMappingStream, line ); ){
+      std::stringstream ss(line);
+      while(ss.good()){
+        std::string substr;
+        std::getline(ss, substr, ' ');
+        result.push_back(substr);
+      }
+      if ((std::stoi(result.at(0))==subdet) && (std::stoi(result.at(1))==layer) && (std::stoi(result.at(2))==moduleU) && (std::stoi(result.at(3))==moduleV)){
+        break;
+      } else{
+        result.clear();
+      }
+    }
+    if(result.empty()){ // for modules not found in the mapping file (currently a few partial modules) use the traditional method.
+      binIDandShares.insert({getTriggerTowerFromEtaPhi(thesum.position().eta(), thesum.position().phi()), 1.0});
+      return binIDandShares;
+    }
+
+    int towerEta;
+    int towerPhi;
+    for (int i=1; i<=std::stoi(result.at(4)); i++){
+      towerEta = 2 + std::stoi(result.at(3*i+2)); // shift by two to avoid negative eta
+      towerPhi = (std::stoi(result.at(3*i+3)) + sector*int(nBinsPhi_)/3 + int(nBinsPhi_)/2) % int(nBinsPhi_);//move to the correct sector
+      if(zside==1){
+        towerPhi = (3*int(nBinsPhi_)/2 - towerPhi - 1) % int(nBinsPhi_); //correct x -> -x in z>0
+      }
+      binIDandShares.insert( {l1t::HGCalTowerID(doNose_, zside, towerEta, towerPhi).rawId(),  std::stod(result.at(3*i+4))/splitDivisor } );
+    }
```

Comment:
All the hardcoded numbers should be implemented as named constants, otherwise it is difficult to understand what they mean exactly